components/brave_wallet/browser/internal/hd_key.cc

```diff
@@ -532,62 +553,75 @@ std::unique_ptr<HDKeyBase> HDKey::DeriveChild(uint32_t index) {
     hdkey->SetPublicKey(public_key);
   }
 
+  if (!path_.empty()) {
+    const std::string node = is_hardened
+                                 ? std::to_string(index - kHardenedOffset) + "'"
+                                 : std::to_string(index);
+
+    hdkey->path_ = path_ + "/" + node;
+  }
   hdkey->depth_ = depth_ + 1;
   hdkey->parent_fingerprint_ = fingerprint_;
   hdkey->index_ = index;
 
-  return std::unique_ptr<HDKeyBase>{hdkey.release()};
+  return hdkey;
 }
 
 std::unique_ptr<HDKeyBase> HDKey::DeriveChildFromPath(const std::string& path) {
-  std::unique_ptr<HDKey> hd_key = std::make_unique<HDKey>();
-  if (path == "m") {
-    if (!private_key_.empty())
-      hd_key->SetPrivateKey(private_key_);
-    else
-      hd_key->SetPublicKey(public_key_);
-    hd_key->chain_code_ = chain_code_;
-    return std::unique_ptr<HDKeyBase>{hd_key.release()};
+  if (path_ != kMasterNode) {
+    LOG(ERROR) << __func__ << ": must derive only from master key";
+    return nullptr;
+  }
+  if (private_key_.empty()) {
+    LOG(ERROR) << __func__ << ": master key must have private key";
+    return nullptr;
   }
+
+  std::unique_ptr<HDKey> hd_key = std::make_unique<HDKey>();
   std::vector<std::string> entries =
       base::SplitString(path, "/", base::WhitespaceHandling::TRIM_WHITESPACE,
                         base::SplitResult::SPLIT_WANT_NONEMPTY);
-  if (entries.empty())
+  if (entries.empty()) {
     return nullptr;
-  for (size_t i = 0; i < entries.size(); ++i) {
+  }
+
+  // Starting with 'm' node and effectively copying `*this` into `hd_key`.
+  if (entries[0] != kMasterNode) {
+    LOG(ERROR) << __func__ << ": path must start with \"m\"";
+    return nullptr;
+  }
+  hd_key->SetPrivateKey(private_key_);
+  hd_key->SetChainCode(chain_code_);
+  hd_key->path_ = path_;
+
+  for (size_t i = 1; i < entries.size(); ++i) {
     std::string entry = entries[i];
-    if (i == 0) {
-      if (entry != "m") {
-        LOG(ERROR) << __func__ << ": path must starts with \"m\"";
-        return nullptr;
-      }
-      if (!private_key_.empty())
-        hd_key->SetPrivateKey(private_key_);
-      else
-        hd_key->SetPublicKey(public_key_);
-      hd_key->chain_code_ = chain_code_;
-      continue;
-    }
```

Comment:
- Strengthened requirements for keys which can derive by path. Must be a master key as I couldn't find any specs on what is a derivation from non master key. Also must have a private key(as it is a master).
- Removed master key handling out of loop so it starts from 1